### PR TITLE
feat: add owner groups to workspace data

### LINKED ISSUE #8560: Return user's groups as a Terraform data source for templates
There may be cases where a template wants to be aware of the user's groups for infrastructure to provision/tags to add, etc.

The easiest way to do it seems like sending them as an array of strings in https://registry.terraform.io/providers/coder/coder/latest/docs/data-sources/workspace

```hcl
data.coder_workspace.me.owner_groups
# ["devops", "platform-engineering", "backstage"]
```

## Use cases

User must be in `devops` group for this to be provisioned as a part of their workspace:
```hcl
resource "special_resource" "for_devops" {
  count = contains(data.coder_workspace.me.owner_groups, "devops") ? 1 : 0
}
```

Add groups as a tag to an resource:

```hcl
resource "aws_instance" "example" {
  tags = {
    Name        = data.coder_workspace.me.name
    OwnerGroups = join(",", data.coder_workspace.me.owner_groups)
  }
}
```

Look up IAM policy based on groups:

```hcl
data "aws_iam_policy" "dynamic" {
  count = length(data.coder_workspace.me.owner_groups)
  arn   = "arn:aws:iam::aws:policy/${element(data.coder_workspace.me.owner_groups, count.index)}"
}
```

### PR DESCRIPTION
**Pull request by f0ssel**
Closes https://github.com/coder/coder/issues/8560

With this change you can now use `coder_workspace.me.owner_groups` in terraform templates and get a list of groups the owner is in. 

As a practical test, I added `MY_GROUPS = jsonencode(data.coder_workspace.me.owner_groups)` to the ENV block of the coder agent and built a workspace:
Screenshot 2024-04-03 at 1 03 46 PM

### CODE REVIEW

File: coderd/provisionerdserver/provisionerdserver.go [hunk 1]
```diff
@@ -467,6 +467,23 @@ func (s *server) acquireProtoJob(ctx context.Context, job database.ProvisionerJo
...
+			if err != nil {
+				return nil, failJob(fmt.Sprintf("get group members: %s", err))
+			}
+			for _, member := range members {
+				if member.ID == owner.ID {
+					ownerGroupNames = append(ownerGroupNames, group.Name)
+					break
+				}
+			}
+		}
```
Comment (sreya) [inline coderd/provisionerdserver/provisionerdserver.go:486]: Wouldn't this be better solved with a query that specifically returns a user's groups?

Comment (f0ssel) [reply coderd/provisionerdserver/provisionerdserver.go:486]: I agree and what I started with, but I saw every other place we get group members we do this pattern. I'll make a new query but I'll probably leave the other places this happens alone for this PR.

Comment (sreya) [reply coderd/provisionerdserver/provisionerdserver.go:486]: Yeah completely fine, I agree the groups code could be improved (something I authored 🙂 )

Comment (f0ssel) [reply coderd/provisionerdserver/provisionerdserver.go]: Using a new query now

**Review by sreya — APPROVED** [review]

File: coderd/database/queries/groups.sql [hunk 2]
```diff
@@ -28,6 +28,29 @@ FROM
...
+    organization_members
+ON
+    organization_members.organization_id = groups.id AND
+    organization_members.user_id = @user_id
+WHERE
+    -- In either case, the group_id will only match an org or a group.
+    (group_members.user_id = @user_id OR organization_members.user_id = @user_id)
+AND
+    -- Ensure the group or organization is the specified organization.
+    groups.organization_id = @organization_id;
```
Comment (Emyrk) [inline coderd/database/queries/groups.sql:51]: I'm guessing this is to cover the "Everyone" group? If so, can you just drop a comment why we join on org_members?

**Review by Emyrk — APPROVED** [review]

File: coderd/database/queries/groups.sql (hunk 2 above)
Comment (f0ssel) [reply coderd/database/queries/groups.sql:53]: Added comments

**Pull request merged**